src/Orleans.Runtime/Placement/Rebalancing/ActivationRebalancerWorker.cs

```diff
@@ -0,0 +1,565 @@
+    }
+
+    private static double ComputeEntropy(
+        Dictionary<SiloAddress, ResourceStatistics>.ValueCollection values,
+        int totalActivations, double meanMemoryUsage)
+    {
+        Debug.Assert(totalActivations > 0);
+        Debug.Assert(meanMemoryUsage > 0);
+
+        var ratios = values.Select(x =>
+            // p_i = (n_i / N) * (m_i / M_m)
+            ((double)x.ActivationCount / totalActivations) * (x.MemoryUsage / meanMemoryUsage));
+
+        var ratiosSum = ratios.Sum();
+        var normalizedRatios = ratios.Select(r => r / ratiosSum);
+
+        const double epsilon = 1e-10d;
+
+        var entropy = -normalizedRatios.Sum(p =>
+        {
+            var value = Math.Max(p, epsilon);  // Avoid log(0)
+            return value * Math.Log(value);    // - sum(p_i * log(p_i))
+        });
+
+        Debug.Assert(entropy > 0);
+
+        return entropy;
+    }
+
+    private static double ComputeHarmonicMean(Dictionary<SiloAddress, ResourceStatistics>.ValueCollection values)
+    {
+        var result = 0d;
+
+        foreach (var value in values)
+        {
+            var count = value.ActivationCount;
+            Debug.Assert(count > 0);
+            result += 1.0 / count;
+        }
+
+        return values.Count / result;
+    }
+
+    private double ComputeAllowedEntropyDeviation(int totalActivations)
+    {
+        const int ActivationThreshold = 10_000;
+        const double MaxAllowedEntropyDeviation = 0.1d;
+
+        if (!_options.ScaleAllowedEntropyDeviation || totalActivations < ActivationThreshold)
+        {
+            return _options.AllowedEntropyDeviation;
+        }
+
+        Debug.Assert(totalActivations > 0);
+
+        var logFactor = (int)Math.Log10(totalActivations / ActivationThreshold);
+        var adjustedDeviation = _options.AllowedEntropyDeviation * Math.Pow(10, logFactor);
+
+        return Math.Min(adjustedDeviation, MaxAllowedEntropyDeviation);
+    }
+
+    private double ComputeAdaptiveScaling(int siloCount, int rebalancingCycle)
+    {
+        Debug.Assert(rebalancingCycle > 0);
+
+        var cycleFactor = 1 - Math.Exp(-_options.CycleNumberWeight * rebalancingCycle);
+        var siloFactor = 1 / (1 + _options.SiloNumberWeight * (siloCount - 1));
+
+        return (double)(cycleFactor * siloFactor);
+    }
+
+    private static List<(SiloAddress, SiloAddress)> FormSiloPairs(
+        Dictionary<SiloAddress, ResourceStatistics> statistics)
+    {
+        var pairs = new List<(SiloAddress, SiloAddress)>();
+        var sorted = statistics.OrderBy(x => x.Value.ActivationCount).ToList();
+
+        var left = 0;
+        var right = sorted.Count - 1;
+
+        while (left < right)
+        {
+            pairs.Add((sorted[left].Key, sorted[right].Key));
+
+            left++;
+            right--;
+        }
+
+        if (left == right) // Odd number of silos
+        {
+            pairs.Add((sorted[left].Key, sorted[left].Key)); // Pair this silo with itself 
+        }
```

Comment:
Yes that would also be acceptable, since it gets ignored by `if (lowSilo.IsSameLogicalSilo(highSilo)) continue`, unless we want to log that we are skipping that silo in this run

Reply:
adjusted